Add the ability to configure disk size and type for GCP clusters (Kube Deploy 638)

spew opened the pull request:
**What this PR does / why we need it**:
This PR enables administrators using gcp-deployer to customize the size and types of the disks in their clusters.

**Special notes for your reviewer**:
Please ignore the first commit, a5bcde4c7236788249e27f9a535c9b073efa2abc (Refactor GCEClient: wrap compute.Service in an interface for mocking GCP compute), it is a prerequisite commit and has its own PR open here: https://github.com/kubernetes-sigs/cluster-api/pull/73.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add the ability to customize the size and type of disks of GCP compute nodes.
```

<!-- All reviews default to cc'ing the kube-deploy-reviewers github group. -->
@kubernetes/kube-deploy-reviewers

krousey commented:
/ok-to-test

krousey commented:
/test all

spew commented:
Rebased on top of the latest.

spew commented:
/retest

spew commented:
/test all

jessicaochen commented on cloud/google/machineactuator.go:
```diff
@@ -709,6 +699,27 @@ func (gce *GCEClient) getImagePath(img string) (imagePath string) {
+	for idx, disk := range config.Disks {
+		diskSizeGb := disk.InitializeParams.DiskSizeGb
+		if diskSizeGb < minDiskSizeGb {
+			diskSizeGb = minDiskSizeGb
```
Might want to spit out a warning log that the configuration is bad and you are defaulting to the min.

spew replied:
Yes, will do.

jessicaochen commented on cloud/google/machineactuator_test.go:
```diff
@@ -0,0 +1,119 @@
+	}
+	return c.mockZoneOperationsGet(project, zone, operation)
+}
+
```
I would suggest creating the cluster and machine in the code instead of parsing yaml files. Parsing is the CLI tool's job and not the machine actuator's job. Plus, it is easier to test more dynamic scenarios when the objects are not hardcoded in a file somewhere.

spew replied:
There's some complexity here in that "machineactuator.go" expects to decode() the raw bytes of machine.Spec.ProviderConfig. For that reason I was trying to use yaml to setup various scenarios as it requires less changes. 

I'm experimenting with some other ideas where the decoding() is decoupled from machineactuator.go. I will likely update the PR with a different approach.

jessicaochen commented on cloud/google/machineactuator_test.go:
```diff
@@ -0,0 +1,119 @@
+	return c.mockZoneOperationsGet(project, zone, operation)
+}
+
+func TestOneDisk(t *testing.T) {
```
I may have just not seen it but is there a test for your logic which will set the min disk size if the given disk size is too small?

spew replied:
Yeah, we should add this test. I will do so. Note that this is pre-existing behavior as of a few days ago where another commit effectively set the disk size to 30GB at all times.

jessicaochen commented on cloud/google/gceproviderconfig/types.go:
```diff
@@ -30,4 +30,15 @@ type GCEProviderConfig struct {
 
 	// The name of the OS to be installed on the machine.
 	OS string `json:"os"`
+	Image       string `json:"image"`
```
The OS is what eventually maps to the OS Image. Why is this field being introduced in this PR?

spew replied:
This was added as an error with resolving a merge conflict. Removing.

jessicaochen commented on cloud/google/gceproviderconfig/types.go:
```diff
@@ -30,4 +30,15 @@ type GCEProviderConfig struct {
+	Disks       []Disk `json:"disks"`
+}
+
+type Disk struct {
```
Since you are modifying the the gce provider config, remember to bump all relevant images after merge. eg. the gce-machine-actuator image.

spew replied:
Thanks for the reminder -- I was not aware of this.

jessicaochen commented on pkg/apis/cluster/v1alpha1/machine_types.go:
```diff
@@ -183,3 +185,16 @@ func (MachineSchemeFns) DefaultingFunction(o interface{}) {
 	log.Printf("Defaulting fields for Machine %s\n", obj.Name)
 }
+
+func ParseMachineYaml(file string) (*Machine, error) {
```
This seems awfully specific to the machine actuator tests. 

It is assuming that the file contains only one machine object and no other objects. Perhaps it is better to co-locate such logic with the tests and then promote to a common location if other code actually uses it?

spew replied:
This will be removed in a future revision of this PR.

spew commented:
After revisiting things to rebase on top of latest as well as attempt to use test fixtures / objects for machine/cluster instead of YAML I reworked things quite a bit. I have a prerequisite commit / PR open here:

   https://github.com/kubernetes-sigs/cluster-api/pull/130

Until that is through, please ignore this PR:

spew commented:
PR 130 is through and committed, so I rebased this commit on top of the latest and it is ready for review again. 

@jessicaochen 
@krousey

jessicaochen commented on cloud/google/machineactuator.go:
```diff
@@ -719,6 +705,28 @@ func (gce *GCEClient) getImagePath(img string) (imagePath string) {
+	for idx, disk := range config.Disks {
+		diskSizeGb := disk.InitializeParams.DiskSizeGb
+		if diskSizeGb < minDiskSizeGb {
+			glog.Info("increasing disk size to %$v gb, the supplied disk size of %v gb is below the minimum", minDiskSizeGb, diskSizeGb)
```
Is the %$v intentional?

jessicaochen replied:
(nit)

spew replied:
Not a nit -- definitely a mistake. Will correct as well as rebase on top of the latest changes (they look significant).

jessicaochen replied:
Ah. In that case, I will withdraw my approval so the PR dashboard will tell me to take a look after the rebase.

spew commented:
PR is rebased onto the tip of master.

spew commented:
@krousey Have you had a chance to look at this before I click merge?

krousey commented:
@spew I've been out for a bit. Looking now. Do no click merge.

krousey commented on cloud/google/gceproviderconfig/v1alpha1/types.go:
```diff
@@ -29,5 +29,15 @@ type GCEProviderConfig struct {
+}
+
+type Disk struct {
+	InitializeParams DiskInitializeParams `json:"initializeParams"`
```
Why add this level of indirection?  Wouldn't
```yaml
disks:
  - diskSizeGb: 30
    diskType: "pd-standard"
  - diskSizeGb: 50
    diskType: "pd-standard"
```

work?

spew replied:
It would be then it would not match the GCE API / structure which is I thought what we wanted to do. It has the InitializeParams structure. 

https://cloud.google.com/compute/docs/reference/rest/beta/instances/insert

Do agree what you are proposing is fine too -- let me know if you want to move to that structure.

krousey commented on cloud/google/machineactuator.go:
```diff
@@ -719,6 +705,28 @@ func (gce *GCEClient) getImagePath(img string) (imagePath string) {
+			AutoDelete: true,
+			Boot:       idx == 0,
+			InitializeParams: &compute.AttachedDiskInitializeParams{
+				SourceImage: imagePath,
```
I think image path is only for the boot disk.

spew replied:
While all disks can have an image, this is a good catch in that the intention of this change was not to put the image on disks other than the boot disk! I'll change this and add a test for this scenario.

krousey commented on cloud/google/gceproviderconfig/v1alpha1/register.go:
```diff
@@ -44,14 +52,62 @@ func addKnownTypes(scheme *runtime.Scheme) error {
+	return &codec, nil
+}
+
+func (codec *GCEProviderConfigCodec) DecodeFromProviderConfig(providerConfig clusterv1.ProviderConfig) (*GCEProviderConfig, error) {
```
cc @mkjelland 
I think we're going to move to where we have separate config for clusters and machines, and also we're going to have provider status for both of them too. Would it be possible for this API to take a `runtime.Object` to decode into instead of returning a specific type?

Also, could `EncodeToProviderConfig` take a `runtime.Object` as well?

This probably doesn't need to be done in this PR.

spew replied:
Happy to take that up in a follow up, more targetted PR.

mkjelland replied:
I can take care of this in my PR to create the separate config types. I was thinking it might be nicer to have 2 methods, that way they can return the correct type. Otherwise the caller has to cast the return value to the type they want, right? Or did you have something else in ming @krousey?

krousey replied:
I was thinking of just having an output parameter of type `runtime.Object`. That way you could do something like

```go
func (codec *GCEProviderConfigCodec) DecodeProviderConfigInto(providerConfig runtime.RawExtension, out runtime.Object) error {...}

var blah GCEProviderConfig
err := c.DecodeProviderConfigInto(m.Spec.ProviderConfig.Value, &blah)
```

krousey commented on cloud/google/machineactuator.go:
```diff
@@ -719,6 +705,28 @@ func (gce *GCEClient) getImagePath(img string) (imagePath string) {
 	return defaultImg
 }
 
+func newDisks(config *gceconfigv1.GCEProviderConfig, zone string, imagePath string, minDiskSizeGb int64) []*compute.AttachedDisk {
```
This min size should eventually be a validation error

spew replied:
Yeah, min size was brought in recent time when we bumped up the min to 30 for custom images, and now for any image (as of a PR a few weeks ago). I was just trying to keep it as it was. I wonder if it should just be removed all together?

krousey commented on cloud/google/machineactuator_test.go:
```diff
@@ -0,0 +1,300 @@
+	checkDiskValues(t, receivedInstance.Disks[1], false, 45, "pd-standard")
+}
+
+func checkInstanceValues(t *testing.T, instance *compute.Instance, diskCount int) {
```
Perhaps make use of https://golang.org/pkg/testing/#T.Helper

spew commented:
Updated post @krousey 's feedback. Noteably:

1. Reworked the "is boot disk" logic. Now only set the image and minimum size if it is the boot disk.
2. Updated the tests to ensure that the image is only set for the boot disk.
3. Started using the Helper() method on Testing.

krousey commented:
/lgtm
